ethcontract-common/src/artifact/hardhat.rs

```diff
@@ -0,0 +1,290 @@
+/// have to filter such networks. See [#545] for more info.
+///
+/// [#545]: https://github.com/gnosis/ethcontract-rs/issues/545.
+pub struct HardHatLoader {
+    /// Override for artifact's origin. If `None`, origin
+    /// will be derived automatically.
+    pub origin: Option<String>,
+
+    /// List of allowed network names and chain IDs.
+    ///
+    /// When loading a contract, networks with names that aren't found
+    /// in this list will be completely ignored. Contracts from these networks
+    /// will not be loaded. You can use this mechanism to bypass
+    /// the requirement that a contract must have the same ABI on all networks.
+    ///
+    /// Empty list means that all networks are allowed.
+    pub networks_allow_list: Vec<String>,
+
+    /// List of denied network names and chain IDs.
+    ///
+    /// When loading a contract, networks with names that are found
+    /// in this list will be completely ignored.
+    ///
+    /// Empty list means that no networks are denied.
+    pub networks_deny_list: Vec<String>,
+}
+
+/// Artifact format.
+#[derive(Copy, Clone, Debug)]
+pub enum Format {
+    /// Contracts for a single network. Generated with `hardhat export`.
+    SingleExport,
+
+    /// Contracts for all networks. Generated with `hardhat export --export-all`.
+    MultiExport,
+}
+
+impl HardHatLoader {
+    /// Create a new hardhat loader.
+    pub fn new() -> Self {
+        HardHatLoader {
+            origin: None,
+            networks_deny_list: Vec::new(),
+            networks_allow_list: Vec::new(),
+        }
+    }
+
+    /// Create a new hardhat loader and set an override for artifact's origins.
+    pub fn with_origin(origin: impl Into<String>) -> Self {
+        HardHatLoader {
+            origin: Some(origin.into()),
+            networks_deny_list: Vec::new(),
+            networks_allow_list: Vec::new(),
+        }
+    }
+
+    /// Set new override for artifact's origin. See [`origin`] for more info.
+    ///
+    /// [`origin`]: #structfield.origin
+    #[inline]
+    pub fn origin(mut self, origin: Option<String>) -> Self {
+        self.origin = origin;
+        self
+    }
+
+    /// Add network name or id to the list of [`allowed networks`].
+    ///
+    /// [`allowed networks`]: #structfield.networks_allow_list
+    #[inline]
+    pub fn allow_network(mut self, network: impl Into<String>) -> Self {
+        self.networks_allow_list.push(network.into());
+        self
+    }
+
+    /// Add network name or id to the list of [`denied networks`].
+    ///
+    /// [`denied networks`]: #structfield.networks_deny_list
+    #[inline]
+    pub fn deny_network(mut self, network: impl Into<String>) -> Self {
+        self.networks_deny_list.push(network.into());
+        self
+    }
+
+    /// Parse a hardhat artifact from JSON string.
+    pub fn load_from_string(&self, format: Format, json: &str) -> Result<Artifact, ArtifactError> {
```

Comment:
I am wondering which of these load_from makes sense. serde_json has https://docs.serde.rs/serde_json/index.html:

- from_reader
- from_slice
- from_str
- from_value

which is slightly different from what we have. Looking at the implementations of those functions it uses https://docs.serde.rs/serde_json/de/trait.Read.html to avoid code duplication. Maybe that would be useful here too.

This is not very important and don't want to sidetrack you too much, just wanted to mention.